docs: combine Blobstream and Blobstream X docs (#1379)
celestiaorg/docs: developers/blobstream.md

description: Learn how to integrate your L2 with Blobstream
---

> # Blobstream: Streaming modular DA to Ethereum

Reviewer: Should this say "Blobstream X" for consistency or just Blobstream?

Reviewer: Just wanted to double-check

Author: https://github.com/celestiaorg/docs/pull/1379#discussion_r1477709788

developers to build high-throughput L2s using Celestia's optimised DA layer,
the first with Data Availability Sampling (DAS).

An implementation of Blobstream, by [Succinct](https://succinct.xyz/), called
[Blobstream X](https://github.com/succinctlabs/blobstreamx), is out

> and will be used in our canonical deployments. This latter proves Celestia 

Reviewer: "nit: "This implementation proves validity of Celestia block headers on a target EVM chain using zero-knowledge proofs (ZKPs), which allows inheriting all the security guarantees of Celestia."

Author: [d7d2f14](https://github.com/celestiaorg/docs/pull/1379/commits/d7d2f14e872ee4fcc5a06079a3e5b0c29840e0f2)

DA require a mechanism for _bridging_ Celestia’s data root to Ethereum as part
of the settlement process. This data root is used during inclusion proofs to
prove that particular rollup transactions were included and made available in
the Celestia network.

Bridging Celestia’s data root to Ethereum requires running a Celestia
_light client_ as a smart contract on Ethereum, to make the latest state
of the Celestia chain known on Ethereum and available to rollups. Blobstream
X utilizes the latest advances in zero-knowledge proofs to generate a
_succinct proof_ that enough Celestia validators have come to consensus
(according to the Tendermint consensus protocol) on a block header, and
verifies this proof in the Blobstream X Ethereum smart contract to update
it with the latest Celestia header.

The Blobstream X zero-knowledge proof not only verifies the consensus of
Celestia validators, but it also merkelizes and hashes all the data roots
in the block range from the previous update to the current update, making
accessible all Celestia data roots (verifiable with a Merkle inclusion proof
against the stored Merkle root) to rollups.

Blobstream X is built and deployed with

> [Succinct’s unified proving stack](https://succinct.xyz/).

Reviewer: can we change this to "Blobstream X is built and deployed with Succinct's protocol" and then link to docs.succinct.xyz? It's currently not up but I should have it up by Monday.

Author: [d7d2f14](https://github.com/celestiaorg/docs/pull/1379/commits/d7d2f14e872ee4fcc5a06079a3e5b0c29840e0f2)

Reviewer 2: leaving this unresolved until docs are up, as a reminder

Reviewer 2: https://github.com/celestiaorg/docs/pull/1379/commits/654b9134117b8fce86c94573f2340d213a55c9c3

![blobstream x draft diagram](/img/blobstream/blobstream_x_draft_diagram.png)

## Integrate with Blobstream X

The following docs go over how developers can integrate Blobstream X.

You can [find the repository for Blobstream X](https://github.com/succinctlabs/blobstreamx)
along with code for:

- The Blobstream X smart contract - [`BlobstreamX.sol`](https://github.com/succinctlabs/blobstreamx/blob/main/contracts/src/BlobstreamX.sol)
- [The Blobstream X circuits](https://alpha.succinct.xyz/celestia/blobstreamx)
- [The Blobstream X contract Golang bindings](https://github.com/succinctlabs/blobstreamx/blob/main/bindings/BlobstreamX.go)

Canonical deployments of Blobstream X will be maintained on the

> following chains: Arbitrum One, Base and Ethereum Mainnet. Every 1

Reviewer: I think on Ethereum Mainnet it will be every 4 hours?

Author: We can change that once we go for the Ethereum mainnet. So far, I think we're focusing on Arbitrum one and Base, which will have 1-hour batches, if I understand correctly. Once we reach mainnet, we can update the documentation to reflect the batches times we're targetting

Reviewer 2: SGTM

can be constructed as `[latestBlock, customTargetBlock)`, with
`latestBlock` is the latest block height that was committed to by the
`BlobstreamX` contract, and `latestBlock > customTargetBlock`,
and `customTargetBlock - latestBlock <= DATA_COMMITMENT_MAX`.

Block ranges that are before the contract's `latestBlock` can't be
proven a second time in different batches.

More information can be found in [`requestHeaderRange(...)`](https://github.com/succinctlabs/blobstreamx/blob/364d3dc8c8dc9fd44b6f9f049cfb18479e56cec4/contracts/src/BlobstreamX.sol#L78-L101)
method.
:::

### How Blobstream X works

As shown in the diagram below, the entrypoint for updates to the Blobstream
X contract is through the `SuccinctGateway` smart contract, which is a
simple entrypoint contract that verifies proofs (against a deployed
onchain verifier for the Blobstream X circuit) and then calls the
`BlobstreamX.sol` contract to update it. More information about the
`SuccinctGateway` can be found here: [TODO].

![blobstream x overview diagram draft](/img/blobstream/blobstream_x_overview_diagram_draft.png)

<!-- markdownlint-disable MD042 -->

:::tip NOTE

> If the contract is still not deployed, then it needs to be

Reviewer: nit: "If the Blobstream X contract is not deployed on a desired chain, it needs to be deployed before it can be used by your rollup. See ..."

Author: [d7d2f14](https://github.com/celestiaorg/docs/pull/1379/commits/d7d2f14e872ee4fcc5a06079a3e5b0c29840e0f2)

deployed before it is used by the prover/relayer. See the
[deployment documentation (TBD)]() for more details.
:::

### How to integrate with Blobstream X

Integrating your L2 with Blobstream X requires two components: your onchain smart
contract logic, and your offchain client logic. The next three sections cover these
topics:

- [Integrate with Blobstream X contracts](./blobstreamx-contracts.md)
- [Integrate with Blobstream X client](./blobstreamx-offchain.md)
- [Querying the Blobstream X proofs](./blobstreamx-proof-queries.md)

### Deployed contracts

You can interact with the Blobstream X contracts today on testnet. The

> Blobstream X Solidity smart contracts are currently deployed on

Reviewer: Should we change these to all just be deployed Blobstream X instances?

Reviewer 2: yes, can you please link them here (once deployed)?

Reviewer 2: I will modify this PR today to:

- [ ] create a section with networks (Arbitrum One, Base, ETH) that we'll launch on, with placeholders for contract addresses

Reviewer 2: https://github.com/celestiaorg/docs/pull/1379/commits/2c6a969a453e10e96e9ed3b1d956a80de8018a4c

Reviewer 2: this is still open to:

- [ ] modify the contract addresses once they have been deployed

Reviewer 2: resolving to bring this comment out

developers to build high-throughput L2s using Celestia's optimised DA layer,
the first with Data Availability Sampling (DAS).

> A new and improved version of Blobstream, Blobstream X, is out and will

Author: We're planning to keep the communications around Blobstream, as a concept, while BlobstreamX being an implementation of Blobstream. This way, end users won't have to know the difference between Blobstream and BlobstreamX etc

> A new and improved version of Blobstream, Blobstream X, is out and will

Reviewer 2: Yes, introduction still introduces the original technology, Blobstream, then talks about the implementation of BlobstreamX as a version of Blobstream.

[Blobstream X](https://github.com/succinctlabs/blobstreamx), is out
and will be used in our canonical deployments. This implementation proves the
validity of Celestia block headers on a target EVM chain using zero-knowledge
proofs which allows, which allow inheriting all the security
guarantees of Celestia.

## What is Blobstream X?

Blobstream X is an implementation of Blobstream with a
ZK light client that bridges Celestia’s modular DA layer to
Ethereum to allow high-throughput rollups to use Celestia’s DA while settling
on Ethereum.

Optimistic or ZK rollups that settle on Ethereum but wish to use Celestia for
DA require a mechanism for _bridging_ Celestia’s data root to Ethereum as part
of the settlement process. This data root is used during inclusion proofs to
prove that particular rollup transactions were included and made available in
the Celestia network.

Bridging Celestia’s data root to Ethereum requires running a Celestia
_light client_ as a smart contract on Ethereum, to make the latest state
of the Celestia chain known on Ethereum and available to rollups. Blobstream
X utilizes the latest advances in zero-knowledge proofs to generate a

> _succinct proof_ that enough Celestia validators have come to consensus

Author: `_succinct proof_` @puma314 can I access to some docs that cover this in depth? I want to learn more about this

Reviewer: tbh we don't have great docs on what's going on--probably the best place to read is just the tendermintx or blobstreamx code

More information can be found in [`requestHeaderRange(...)`](https://github.com/succinctlabs/blobstreamx/blob/364d3dc8c8dc9fd44b6f9f049cfb18479e56cec4/contracts/src/BlobstreamX.sol#L78-L101)
method.
:::

### How Blobstream X works

As shown in the diagram below, the entrypoint for updates to the Blobstream
X contract is through the `SuccinctGateway` smart contract, which is a
simple entrypoint contract that verifies proofs (against a deployed
onchain verifier for the Blobstream X circuit) and then calls the
`BlobstreamX.sol` contract to update it.
[Find more information about the `SuccinctGateway`](https://docs.succinct.xyz/platform/onchain-integration#succinct-gateway).

![blobstream x overview diagram draft](/img/blobstream/Celestia_Blobstream_X2b.png)

<!-- markdownlint-disable MD042 -->

:::tip NOTE
If the Blobstream X contract is not deployed on a desired chain,
it needs to be deployed before it can be used by your rollup. See the
[deployment documentation](https://docs.succinct.xyz/platform/onchain-integration#gateway-deployment)
for more details.
:::

### How to integrate with Blobstream X

Integrating your L2 with Blobstream X requires two components: your onchain smart

> contract logic, and your offchain client logic. The next three sections cover these

Author: `what's the offchain client logic`? is it access to an RPC? if so, then I prefer stating that clearly

Reviewer 2: it will be covered in here https://docs.celestia.org/developers/blobstream-offchain. should i change the path of the page back to `blobstream-offchain` instead of `blobstream-x-offchain` @rach-id ?

Reviewer 2: and other pages: 
/developers/blobstream-contracts
/developers/blobstream-offchain
/developers/blobstream-proof-queries

Author: thanks 🙏

Reviewer 2: https://github.com/celestiaorg/docs/pull/1379/commits/31360c21ad0b92e50f12d04396d28f471e0c1cdc

<!-- markdownlint-disable MD013 -->

| Contract     | EVM network      | Contract address                                                                                                                | Attested data on Celestia |
| ------------ | ---------------- | ------------------------------------------------------------------------------------------------------------------------------- | ------------- |
| Blobstream X  | Ethereum Mainnet          | [`Not yet deployed`](https://etherscan.io/address/0xTODO) | [Mainnet Beta](../nodes/mainnet.md) |
| Blobstream X | Arbitrum One | [`Not yet deployed`](https://arbiscan.io/address/0xTODO)  | [Mainnet Beta](../nodes/mainnet.md) |
| Blobstream X | Base           | [`Not yet deployed`](https://goerli.etherscan.io/address/0xTODO)  | [Mainnet Beta](../nodes/mainnet.md) |
| Blobstream X | Ethereum Sepolia           | [`Not yet deployed`](https://sepolia.etherscan.io/address/0xTODO)  | [Mocha testnet](../nodes/mocha-testnet.md) |
| Blobstream X | Arbitrum Sepolia           | [`Not yet deployed`](https://sepolia.arbiscan.io/address/0xTODO)  | [Mocha testnet](../nodes/mocha-testnet.md) |

<!-- markdownlint-enable MD013 -->

> ## Blobstream X vs. data availability committees (DACs)

Author: ```suggestion
## Blobstream vs. data availability committees (DACs)
```

This should stay as Blobstream because we are discussing the concept of Blobstream, not the implementation

### Decentralization and security

> Blobstream X is built on Celestia, which uses a CometBFT-based proof-of-stake

Author: ```suggestion
Blobstream is built on Celestia, which uses a CometBFT-based proof-of-stake
```

> system. Blobstream X shares the same security assumptions

Author: ```suggestion
system. Blobstream shares the same security assumptions
```

as Celestia. In contrast, data availability committees (DACs), are typically
centralized or semi-centralized, relying on a specific set of entities or
individuals to vouch for data availability.

### Mechanism of verification

> Blobstream X uses data availability attestations, which are Merkle roots of

Author: ```suggestion
Blobstream uses data availability attestations, which are Merkle roots of
```

the batched L2 data, to confirm that the necessary data is present on Celestia.

> The L2 contract on Ethereum can check directly with Blobstream X if the data

Author: ```suggestion
The L2 contract on Ethereum can check directly with Blobstream if the data
```

is published on Celestia. Similarly, a DAC would rely on
attestations or confirmations from its permissioned members.

### Flexibility and scalability

> Blobstream X is designed to offer high-throughput data availability for Ethereum

Author: ```suggestion
Blobstream is designed to offer high-throughput data availability for Ethereum
```

L2s, aiming to strike a balance between scalability and security. It operates
independently of Ethereum's gas costs, as Celestia's resource pricing is more
byte-focused rather than computation-centric. On the other hand, the scalability
and flexibility of a DAC would depend on its specific design and implementation.

> In summary, both Blobstream X and DACs aim to ensure offchain data availability,

Author: ```suggestion
In summary, both Blobstream and DACs aim to ensure offchain data availability,
```

> but Blobstream X offers a more decentralized, secure, and scalable solution

Author: ```suggestion
but Blobstream offers a more decentralized, secure, and scalable solution
```

developers to build high-throughput L2s using Celestia's optimised DA layer,
the first with Data Availability Sampling (DAS).

An implementation of Blobstream, by [Succinct](https://docs.succinct.xyz/), called
[Blobstream X](https://github.com/succinctlabs/blobstreamx), is out

> and will be used in our canonical deployments. This implementation proves the

Reviewer 2: ```suggestion
and will be used in the canonical deployments. This implementation proves the
```

> validity of Celestia block headers on a target EVM chain using zero-knowledge

Reviewer 2: ```suggestion
validity of Celestia block headers on a target EVM chain using zero-knowledge (ZK)
```

proofs, which allow inheriting all the security
guarantees of Celestia.

## What is Blobstream X?

Blobstream X is an implementation of Blobstream with a

> ZK light client that bridges Celestia’s modular DA layer to

Reviewer 2: ```suggestion
zero knowledge (ZK) light client that bridges Celestia’s modular DA layer to
```

Ethereum to allow high-throughput rollups to use Celestia’s DA while settling
on Ethereum.

> Optimistic or ZK rollups that settle on Ethereum but wish to use Celestia for

Reviewer 2: ```suggestion
Optimistic or ZK rollups that settle on Ethereum, but wish to use Celestia for
```

> DA require a mechanism for _bridging_ Celestia’s data root to Ethereum as part

Reviewer 2: ```suggestion
DA, require a mechanism for _bridging_ Celestia’s data root to Ethereum as part
```

of the settlement process. This data root is used during inclusion proofs to
prove that particular rollup transactions were included and made available in
the Celestia network.

Bridging Celestia’s data root to Ethereum requires running a Celestia
_light client_ as a smart contract on Ethereum, to make the latest state
of the Celestia chain known on Ethereum and available to rollups. Blobstream

> X utilizes the latest advances in zero-knowledge proofs to generate a

Reviewer 2: ```suggestion
X utilizes the latest advances in ZK proofs to generate a
```

_succinct proof_ that enough Celestia validators have come to consensus

> (according to the Tendermint consensus protocol) on a block header, and

Reviewer 2: ```suggestion
(according to the CometBFT consensus protocol) on a block header, and
```

> (according to the Tendermint consensus protocol) on a block header, and

Reviewer 2: In the section further down the page "Blobstream vs. DACs" there is a reference to CometBFT. What should we use as standard in the docs @rach-id ?

Author: cometBFT

verifies this proof in the Blobstream X Ethereum smart contract to update
it with the latest Celestia header.

> The Blobstream X zero-knowledge proof not only verifies the consensus of

Reviewer 2: ```suggestion
The Blobstream X ZK proof not only verifies the consensus of
```

Celestia validators, but it also merkelizes and hashes all the data roots
in the block range from the previous update to the current update, making
accessible all Celestia data roots (verifiable with a Merkle inclusion proof
against the stored Merkle root) to rollups.

Blobstream X is built and deployed with
[Succinct's protocol](https://docs.succinct.xyz).

![blobstream x draft diagram](/img/blobstream/Celestia_Blobstream_X1b.png)

## Integrate with Blobstream X

The following docs go over how developers can integrate Blobstream X.

You can [find the repository for Blobstream X](https://github.com/succinctlabs/blobstreamx)
along with code for:

> - The Blobstream X smart contract - [`BlobstreamX.sol`](https://github.com/succinctlabs/blobstreamx/blob/main/contracts/src/BlobstreamX.sol)

Reviewer 2: ```suggestion
- [The Blobstream X smart contract - `BlobstreamX.sol`](https://github.com/succinctlabs/blobstreamx/blob/main/contracts/src/BlobstreamX.sol)
```

- [The Blobstream X circuits](https://alpha.succinct.xyz/celestia/blobstreamx)
- [The Blobstream X contract Golang bindings](https://github.com/succinctlabs/blobstreamx/blob/main/bindings/BlobstreamX.go)

Canonical deployments of Blobstream X will be maintained on the
following chains: Arbitrum One, Base and Ethereum Mainnet. Every 1
hour, Succinct will post an update to the Blobstream X contract
that will include a new data commitment range that covers a 1-hour
block range from the `latestBlock` in the Blobstream X contract.
On Ethereum Mainnet, the Blobstream X contract will be updated
every 4 hours.

:::tip NOTE
Custom ranges can be requested using the `BlobstreamX` contract
to create proofs for specific Celestia block batches. These ranges
can be constructed as `[latestBlock, customTargetBlock)`, with
`latestBlock` is the latest block height that was committed to by the
`BlobstreamX` contract, and `latestBlock > customTargetBlock`,
and `customTargetBlock - latestBlock <= DATA_COMMITMENT_MAX`.

Block ranges that are before the contract's `latestBlock` can't be
proven a second time in different batches.

> More information can be found in [`requestHeaderRange(...)`](https://github.com/succinctlabs/blobstreamx/blob/364d3dc8c8dc9fd44b6f9f049cfb18479e56cec4/contracts/src/BlobstreamX.sol#L78-L101)

Reviewer 2: ```suggestion
More information can be found in the [`requestHeaderRange(...)`](https://github.com/succinctlabs/blobstreamx/blob/364d3dc8c8dc9fd44b6f9f049cfb18479e56cec4/contracts/src/BlobstreamX.sol#L78-L101)
```

target EVM chain using zk-proofs which allows inheriting all the security
guarantees of Celestia.

## Blobstream X

![blobstream x](/img/blobstream/blobstream_x.png)

BlobstreamX is a ZK light client that bridges Celestia’s modular DA layer to
Ethereum to allow high-throughput rollups to use Celestia’s DA while settling
on Ethereum.

Optimistic or ZK rollups that settle on Ethereum but wish to use Celestia for
DA require a mechanism for _bridging_ Celestia’s data root to Ethereum as part

> of the settlement process. This data root is used during validity proofs to

Author: probably: `during validity proofs`-> `during inclusion proofs`

accessible all Celestia data roots (verifiable with a Merkle inclusion proof
against the stored Merkle root) to rollups.

Blobstream X is built and deployed with
[Succinct’s unified proving stack](https://succinct.xyz/).

![blobstream x draft diagram](/img/blobstream/blobstream_x_draft_diagram.png)

## Integrate with Blobstream X

The following docs go over how developers can integrate Blobstream X.

### Overview

You can [find the repository for Blobstream X](https://github.com/succinctlabs/blobstreamx)
along with code for:

- The Blobstream X smart contract - [`BlobstreamX.sol`](https://github.com/succinctlabs/blobstreamx/blob/main/contracts/src/BlobstreamX.sol)

> - [The Blobstream X circuits](https://alpha.succinct.xyz/celestia/blobstreamx)

Author: probably also add links to the golang binding : https://github.com/succinctlabs/blobstreamx/tree/main/bindings

Canonical deployments of Blobstream X will be maintained on the
following chains: Arbitrum One, Base and Ethereum Mainnet. Every 4

> hours, Succinct will post an update to the Blobstream X contract

Author: I guess 1h or so

> that will include a new data commitment range that covers a 4-hour

Author: 1hour
